pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,698 @@
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageosMounter
+	pvUID types.UID
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageosMounter
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	adminSecretName := ""
+	adminSecretNamespace := "default"
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "adminsecretname":
+			adminSecretName = v
+		case "adminsecretnamespace":
+			adminSecretNamespace = v
+		case "volumenamespace":
+			c.volNamespace = v
+		case "description":
+			c.description = v
+		case "pool":
+			c.pool = v
+		case "fstype":
+			c.fsType = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+		}
+	}
+
+	// Set from PVC
+	c.podNamespace = c.options.PVC.Namespace
+	c.volName = c.options.PVName
+	if c.volNamespace == "" {
+		c.volNamespace = c.options.PVC.Namespace
+	}
+	c.labels = make(map[string]string)
+	for k, v := range c.options.PVC.Labels {
+		c.labels[k] = v
+	}
+	capacity := c.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	c.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	if adminSecretName == "" {
+		return nil, fmt.Errorf("missing StorageOS admin secret name")
+	}
+
+	apiCfg, err := parsePVSecret(adminSecretNamespace, adminSecretName, c.plugin.host.GetKubeClient())
+	if err != nil {
+		return nil, err
+	}
+	c.apiCfg = apiCfg
+
+	vol, err := c.manager.CreateVolume(c)
+	if err != nil {
+		glog.Errorf("failed to create volume: %v", err)
+		return nil, err
+	}
+	if vol.FSType == "" {
+		vol.FSType = defaultFSType
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   vol.Name,
+			Labels: map[string]string{},
+			Annotations: map[string]string{
+				volumehelper.VolumeDynamicallyCreatedByKey: "storageos-dynamic-provisioner",
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: c.options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   c.options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", vol.Size)),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				StorageOS: &v1.StorageOSVolumeSource{
+					VolumeName:      vol.Name,
+					VolumeNamespace: vol.Namespace,
+					FSType:          vol.FSType,
+					ReadOnly:        false,
+					SecretRef: &v1.LocalObjectReference{
+						Name: adminSecretName,
```

Comment:
The other volume plugins that make use of a secret name/namespace in the storage class parameters only do so during the Provision()/Delete() operations, which are run by the persistent volume controller, which can look up the storage class and any secret.

This plugin expects to have access to that secret from the node as part of the mount/unmount operation.

The adminSecretNamespace specified in the storage class is lost here...
1. that means the node mount/unmount operation won't know what namespace to look in to find the secret referenced by the storage class
2. even if it knew where to look, it would be forbidden (when running with the Node authorizer enabled)

For the node to have the info required to locate the secret, and the authorizer to allow the node to access the secret, I think the PV volume source has to allow specifying the secret name and namespace (which means the PV storageos volume source looks different than the Pod storageos volume source which only allows a local reference).

Then the mount/unmount operation has the info needed to locate the secret, and the node authorizer has the info needed to permit access to it.